### ISSUE TITLE
About holding stock minutes

### ISSUE THREAD

**Issue by kroutony** [woocommerce/woocommerce#24581]
"Hold stock (for unpaid orders) for x minutes. When this limit is reached,  pending order will be cancelled. Leave blank to disable."
When I read this line, I thought it's an option to hold stock in x minutes for every unpaid order.
=> Unpaid order will be cancel after x minutes from the time being created.
But after I reviewing the source code, I realize it's not like that. I found the system will check all unpaid orders that its creation time before x minutes from now in "every x minutes".
Basically The timing to process a certain unpaid order is not at the time "order time + x minutes". It will be the schedule time. The order will be cancelled between "order time + x minutes" and "order time + 2x minutes"

**Comment by vedanshujain** [comment]
Yes, you are correct. The order will indeed be canceled between "order time + x minutes" and "order time + 2x minutes" as you say.

Maybe it will be useful if we have a filter while scheduling the cron so that people can run cron more frequently if needed.

**Comment by mfs-mindsize** [comment]
What's the status of this? It looks like there might have been a PR? Or not?

**Comment by mfs-mindsize** [comment]
@kroutony - I had some time to dig into this. As you noted, the "Hold stock (for unpaid orders) for x minutes" does not set an age/delay at the order level, but instead that value is the interval between checking for orders to cancel. 

All is not lost :)

You probably noticed, that interval is set using wp_schedule_single_event().

[https://developer.wordpress.org/reference/functions/wp_schedule_single_event/](https://developer.wordpress.org/reference/functions/wp_schedule_single_event/)

This function has a filter (on line 80):

`$pre = apply_filters( 'pre_schedule_event', null, $event );`

In the short term, you can tap into that filter, hijack the interval-based scheduling of woocommerce_cancel_unpaid_orders, and schedule woocommerce_cancel_unpaid_orders yourself with a more appropriate interval for your needs. 

In the longer term, I'm looking into a WC core-based solution, with a proper pull request and so on. The actual work could be done sooner, but the process (read: time) for merging into core is an unknown. So I wanted to let you know there is in fact a short-term solution, if you're willing and able.

**Comment by mfs-mindsize** [comment]
@kroutony - Quick follow up. 

I opted or a clean and quick solution. I've "decoupled" the order "age" from the wp_schedule_single_event()'s $timestamp. The $timestamp for the hook: woocommerce_cancel_unpaid_orders has a default of 5 (minutes) and that value can be updated using the new filter: 'woocommerce_cancel_unpaid_orders_interval_minutes'. 

When the PR is submitted (after an internal review on my end), reviewed by WC, and so on IDK.

**Comment by mfs-mindsize** [comment]
FYI - PR submitted: https://github.com/woocommerce/woocommerce/pull/28933

**Comment by roykho** [comment]
Closed via https://github.com/woocommerce/woocommerce/pull/28933
